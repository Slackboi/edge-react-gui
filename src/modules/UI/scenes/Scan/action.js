// @flow

import { Alert } from 'react-native'
import { Actions } from 'react-native-router-flux'
import type { EdgeParsedUri, EdgeSpendInfo, EdgePaymentProtocolInfo } from 'edge-core-js'

import * as Constants from '../../../../constants/indexConstants.js'
import type { Dispatch, GetState } from '../../../ReduxTypes.js'
import { getPaymentProtocolInfo, parseUri } from '../../../Core/Wallets/api.js'
import { isEdgeLogin, denominationToDecimalPlaces, noOp } from '../../../utils.js'
import { loginWithEdge } from '../../../../actions/EdgeLoginActions.js'
import { spendRequested, paymentProtocolSpendRequested } from '../SendConfirmation/action.js'
import s from '../../../../locales/strings.js'

import { activated as legacyAddressModalActivated, deactivated as legacyAddressModalDeactivated } from './LegacyAddressModal/LegacyAddressModalActions.js'
import { activated as privateKeyModalActivated } from './PrivateKeyModal/PrivateKeyModalActions.js'

export const PREFIX = 'SCAN/'

export const UPDATE_RECIPIENT_ADDRESS = 'UPDATE_RECIPIENT_ADDRESS'

export const TOGGLE_ENABLE_TORCH = 'TOGGLE_ENABLE_TORCH'
export const toggleEnableTorch = () => ({
  type: TOGGLE_ENABLE_TORCH
})

export const TOGGLE_ADDRESS_MODAL_VISIBILITY = 'TOGGLE_ADDRESS_MODAL_VISIBILITY'
export const toggleAddressModal = () => ({
  type: TOGGLE_ADDRESS_MODAL_VISIBILITY
})

export const ENABLE_SCAN = 'ENABLE_SCAN'
export const enableScan = () => {
  return {
    type: ENABLE_SCAN
  }
}

export const DISABLE_SCAN = 'DISABLE_SCAN'
export const disableScan = () => {
  return {
    type: DISABLE_SCAN
  }
}

export const PARSE_URI_SUCCEEDED = 'PARSE_URI_SUCCEEDED'
export const parseUriSucceeded = (parsedUri: EdgeParsedUri) => ({
  type: PARSE_URI_SUCCEEDED,
  data: { parsedUri }
})

export const PARSE_URI_FAILED = 'PARSE_URI_FAILED'
export const parseUriFailed = (error: Error) => ({
  type: PARSE_URI_FAILED,
  data: { error }
})

export const PARSE_URI_RESET = 'PARSE_URI_RESET'
export const parseUriReset = () => ({
  type: PARSE_URI_RESET
})

export const parseUriRequested = (data: string) => (dispatch: Dispatch, getState: GetState) => {
  if (!data) return
  const state = getState()
  const selectedWalletId = state.ui.wallets.selectedWalletId
  const edgeWallet = state.core.wallets.byId[selectedWalletId]
  const guiWallet = state.ui.wallets.byId[selectedWalletId]
  if (isEdgeLogin(data)) {
    // EDGE LOGIN
    dispatch(loginWithEdge(data))
    Actions[Constants.EDGE_LOGIN]()
    return
  }

  parseUri(edgeWallet, data).then(
    (parsedUri: EdgeParsedUri) => {
      dispatch(parseUriSucceeded(parsedUri))

      if (parsedUri.token) {
        // TOKEN URI
        const { contractAddress, currencyName, multiplier } = parsedUri.token
        const currencyCode = parsedUri.token.currencyCode.toUpperCase()
        let decimalPlaces = 18
        if (parsedUri.token && parsedUri.token.multiplier) {
          decimalPlaces = denominationToDecimalPlaces(parsedUri.token.multiplier)
        }
        const parameters = {
          contractAddress,
          currencyCode,
          currencyName,
          multiplier,
          decimalPlaces,
          walletId: selectedWalletId,
          wallet: guiWallet,
          onAddToken: noOp
        }
        return Actions.addToken(parameters)
      }

      if (isLegacyAddressUri(parsedUri)) {
        // LEGACY ADDRESS URI
        return dispatch(legacyAddressModalActivated())
      }

      if (isPrivateKeyUri(parsedUri)) {
        // PRIVATE KEY URI
        return dispatch(privateKeyModalActivated())
      }

      if (isPaymentProtocolUri(parsedUri)) {
        // BIP70 URI
        return Promise.resolve(parsedUri.paymentProtocolURL)
          .then(paymentProtocolURL => getPaymentProtocolInfo(edgeWallet, paymentProtocolURL))
          .then(paymentProtocolInfo => convertPaymentProtocolInfoToSpendInfo(paymentProtocolInfo))
          .then(spendInfo => {
            dispatch(paymentProtocolSpendRequested(spendInfo))
            return Actions.sendConfirmation('fromScan')
          })
      }

      // PUBLIC ADDRESS URI
      const spendInfo = convertParsedUriToSpendInfo(parsedUri)
      dispatch(spendRequested(spendInfo))
      return Actions.sendConfirmation('fromScan')
    },
    () => {
      // INVALID URI
      dispatch(disableScan())
      setTimeout(
        () =>
          Alert.alert(s.strings.scan_invalid_address_error_title, s.strings.scan_invalid_address_error_description, [
            { text: s.strings.string_ok, onPress: () => dispatch(enableScan()) }
          ]),
        500
      )
    }
  )
}

export const qrCodeScanned = (data: string) => (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const isScanEnabled = state.ui.scenes.scan.scanEnabled
  if (!isScanEnabled) return

  dispatch(disableScan())
  dispatch(parseUriRequested(data))
}

export const addressModalDoneButtonPressed = (data: string) => (dispatch: Dispatch, getState: GetState) => {
  dispatch(parseUriRequested(data))
}

export const addressModalCancelButtonPressed = () => (dispatch: Dispatch, getState: GetState) => {
  // dispatch(addressModalDeactivated())
}

export const legacyAddressModalContinueButtonPressed = () => (dispatch: Dispatch, getState: GetState) => {
  dispatch(legacyAddressModalDeactivated())
  dispatch(enableScan())

  const state = getState()
  const parsedUri = state.ui.scenes.scan.parsedUri
  if (!parsedUri) return

  const spendInfo = convertParsedUriToSpendInfo(parsedUri)
  dispatch(spendRequested(spendInfo))
  Actions.sendConfirmation('fromScan')
}

export const legacyAddressModalCancelButtonPressed = () => (dispatch: Dispatch) => {
  dispatch(legacyAddressModalDeactivated())
  dispatch(enableScan())
}

export const isTokenUri = (parsedUri: EdgeParsedUri): boolean => {
  return !!parsedUri.token
}

export const isLegacyAddressUri = (parsedUri: EdgeParsedUri): boolean => {
  return !!parsedUri.legacyAddress
}

export const isPrivateKeyUri = (parsedUri: EdgeParsedUri): boolean => {
  return !!parsedUri.privateKeys && parsedUri.privateKeys.length >= 1
}

export const isPaymentProtocolUri = (parsedUri: EdgeParsedUri): boolean => {
  return !!parsedUri.paymentProtocolURL && !parsedUri.publicAddress
}

export const convertParsedUriToSpendInfo = (parsedUri: EdgeParsedUri): EdgeSpendInfo => ({
  currencyCode: parsedUri.currencyCode,
  spendTargets: [
    {
      publicAddress: parsedUri.publicAddress,
      nativeAmount: parsedUri.nativeAmount,
      otherParams: { uniqueIdentifier: parsedUri.uniqueIdentifier }
    }
  ],
  metadata: parsedUri.metadata
})

const BITPAY = {
  domain: 'bitpay.com',
  merchantName: (memo: string) => {
    // Example BitPay memo
    // "Payment request for BitPay invoice DKffym7WxX6kzJ73yfYS7s for merchant Electronic Frontier Foundation"
    // eslint-disable-next-line no-unused-vars
    const [_, merchantName] = memo.split(' for merchant ')
    return merchantName
  }
}

export const convertPaymentProtocolInfoToSpendInfo = (paymentProtocolInfo: EdgePaymentProtocolInfo): Promise<EdgeSpendInfo> => {
  const { domain, memo, merchant, nativeAmount, spendTargets } = paymentProtocolInfo

  const name = domain === BITPAY.domain ? BITPAY.merchantName(memo) : merchant || domain
  const notes = memo

  return Promise.resolve({
    networkFeeOption: 'standard',
    metadata: {
      name,
      notes
    },
    nativeAmount,
    spendTargets: spendTargets.map(target => (target.otherParams ? target : { ...target, otherParams: {} }))
  })
}
