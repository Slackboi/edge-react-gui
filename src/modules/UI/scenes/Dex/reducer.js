// @flow

import { UPDATE_TOKEN_LIST } from './action.js'
import { combineReducers } from 'redux'
import type { DirectoryTokenInfo } from '../../../../types.js'
import type { Action } from '../../../ReduxTypes.js'

export const tokenDirectory = (state: Array<DirectoryTokenInfo> = [], action: Action) => {
  switch (action.type) {
    case UPDATE_TOKEN_LIST:
      return action.data
    default:
      return state
  }
}

export const dex = combineReducers({
  tokenDirectory
})
