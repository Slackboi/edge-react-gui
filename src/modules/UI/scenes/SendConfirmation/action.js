// @flow

import { bns } from 'biggystring'
import type { EdgeMetadata, EdgeParsedUri, EdgeTransaction, EdgeSpendInfo } from 'edge-core-js'
import { Alert } from 'react-native'
import { Actions } from 'react-native-router-flux'

import { OPEN_AB_ALERT } from '../../../../constants/indexConstants'
import { getWallet } from '../../../Core/selectors.js'
import {
  broadcastTransaction,
  getMaxSpendable,
  makeSpend,
  saveTransaction,
  signTransaction,
  getPaymentProtocolInfo,
  makeSpendInfo
} from '../../../Core/Wallets/api.js'
import type { Dispatch, GetState } from '../../../ReduxTypes'
import { openABAlert } from '../../components/ABAlert/action'
import { getSelectedWalletId } from '../../selectors.js'
import { getSpendInfo, getTransaction } from './selectors'
import type { GuiMakeSpendInfo } from './selectors'
import { getAccount } from '../../../Core/selectors.js'
import { checkPin } from '../../../Core/Account/api.js'
import { getExchangeDenomination } from '../../selectors.js'

import s from '../../../../locales/strings.js'

type EdgePaymentProtocolUri = EdgeParsedUri & { paymentProtocolURL: string }

const PREFIX = 'UI/SendConfimation/'
export const UPDATE_IS_KEYBOARD_VISIBLE = PREFIX + 'UPDATE_IS_KEYBOARD_VISIBLE'
export const UPDATE_SPEND_PENDING = PREFIX + 'UPDATE_SPEND_PENDING'
export const RESET = PREFIX + 'RESET'
export const UPDATE_PAYMENT_PROTOCOL_TRANSACTION = PREFIX + 'UPDATE_PAYMENT_PROTOCOL_TRANSACTION'
export const UPDATE_TRANSACTION = PREFIX + 'UPDATE_TRANSACTION'

export const updateAmount = (nativeAmount: string, exchangeAmount: string, fiatPerCrypto: string) => (dispatch: Dispatch, getState: GetState) => {
  const amountFiatString: string = bns.mul(exchangeAmount, fiatPerCrypto)
  const amountFiat: number = parseFloat(amountFiatString)
  const metadata: EdgeMetadata = { amountFiat }
  dispatch(createTX({ nativeAmount, metadata }, false))
}

export const paymentProtocolUriReceived = ({ paymentProtocolURL }: EdgePaymentProtocolUri) => (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const walletId = getSelectedWalletId(state)
  const edgeWallet = getWallet(state, walletId)

  Promise.resolve(paymentProtocolURL)
    .then(paymentProtocolURL => getPaymentProtocolInfo(edgeWallet, paymentProtocolURL))
    .then(makeSpendInfo)
    .then(spendInfo => {
      dispatch(newSpendInfo(spendInfo))

      return makeSpend(edgeWallet, spendInfo).then(
        edgeTransaction => {
          dispatch(updatePaymentProtocolTransaction(edgeTransaction))
          Actions.sendConfirmation('fromScan')
        },
        error => {
          dispatch(makeSpendFailed(error))
          Actions.sendConfirmation('fromScan')
        }
      )
    })
    .catch((error: Error) => {
      console.log(error)
      setTimeout(
        () => Alert.alert(s.strings.scan_invalid_address_error_title, s.strings.scan_invalid_address_error_description, [{ text: s.strings.string_ok }]),
        500
      )
    })
}

export const createTX = (parsedUri: GuiMakeSpendInfo | EdgeParsedUri, forceUpdateGui?: boolean = true) => (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const walletId = getSelectedWalletId(state)
  const edgeWallet = getWallet(state, walletId)
  const parsedUriClone = { ...parsedUri }
  const spendInfo = getSpendInfo(state, parsedUriClone)

  makeSpend(edgeWallet, spendInfo)
    .then(edgeTransaction => {
      dispatch(updateTransaction(edgeTransaction, parsedUriClone, forceUpdateGui, null))
    })
    .catch(e => {
      dispatch(updateTransaction(null, parsedUriClone, forceUpdateGui, e))
    })
}

export const updateMaxSpend = () => (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const walletId = getSelectedWalletId(state)
  const edgeWallet = getWallet(state, walletId)
  const spendInfo = getSpendInfo(state)
  dispatch(newSpendInfo(spendInfo))
  dispatch(confirmSpendingLimits())

  getMaxSpendable(edgeWallet, spendInfo)
    .then(nativeAmount => {
      dispatch(createTX({ nativeAmount }, true))
    })
    .catch(e => {
      console.log(e)
    })
}

export const signBroadcastAndSave = () => async (dispatch: Dispatch, getState: GetState) => {
  dispatch(updateSpendPending(true))
  const state = getState()
  const selectedWalletId = getSelectedWalletId(state)
  const wallet = getWallet(state, selectedWalletId)
  const edgeUnsignedTransaction = getTransaction(state)
  let edgeSignedTransaction = edgeUnsignedTransaction

  try {
    edgeSignedTransaction = await signTransaction(wallet, edgeUnsignedTransaction)
    edgeSignedTransaction = await broadcastTransaction(wallet, edgeSignedTransaction)
    await saveTransaction(wallet, edgeSignedTransaction)
    dispatch(updateSpendPending(false))
    Actions.pop()
    const successInfo = {
      success: true,
      title: 'Transaction Sent',
      message: 'Your transaction has been successfully sent.'
    }
    dispatch(openABAlert(OPEN_AB_ALERT, successInfo))
  } catch (e) {
    dispatch(updateSpendPending(false))
    const errorInfo = {
      success: false,
      title: 'Transaction Failure',
      message: e.message
    }
    dispatch(updateTransaction(edgeSignedTransaction, null, true, new Error('broadcastError')))
    dispatch(openABAlert(OPEN_AB_ALERT, errorInfo))
  }
}

export const reset = () => ({
  type: RESET,
  data: {}
})

export const updatePaymentProtocolTransaction = (transaction: EdgeTransaction) => ({
  type: UPDATE_PAYMENT_PROTOCOL_TRANSACTION,
  data: { transaction }
})

export const updateTransaction = (
  transaction: ?EdgeTransaction,
  parsedUri: ?EdgeParsedUri,
  forceUpdateGui: ?boolean,
  error: ?Error,
  pinIsEnabled: boolean,
  pinAmount: number
) => ({
  type: UPDATE_TRANSACTION,
  data: { transaction, parsedUri, forceUpdateGui, error, pinIsEnabled, pinAmount }
})

export const updateSpendPending = (pending: boolean) => ({
  type: UPDATE_SPEND_PENDING,
  data: { pending }
})

export const INCORRECT_PIN = PREFIX + 'INCORRECT_PIN'
export const incorrectPin = () => ({
  type: INCORRECT_PIN,
  data: {}
})

export const MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED = PREFIX + 'MAKE_SPEND_FAILED'
export const makeSpendFailed = (error: Error) => ({
  type: MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED,
  data: { error }
})

export const NEW_SPEND_INFO = PREFIX + 'NEW_SPEND_INFO'
export const newSpendInfo = (spendInfo: EdgeSpendInfo) => ({
  type: NEW_SPEND_INFO,
  data: { spendInfo }
})

export const PIN_CHANGED = PREFIX + 'PIN_CHANGED'
export const pinChanged = (pin: string) => ({
  type: PIN_CHANGED,
  data: { pin }
})

export { createTX as updateMiningFees, createTX as updateParsedURI, createTX as uniqueIdentifierUpdated }

const confirmSpendingLimits = () => (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const account = getAccount(state)
  const { convertCurrency } = getCurrencyConverter(state)
  const { isEnabled: pinIsEnabled, amount: pinAmount } = state.ui.settings.spendingLimits.transaction
  const spendInfo = state.ui.sendConfirmation.spendInfo
  const { nativeAmount, currencyCode } = spendInfo
  const isoFiatCurrencyCode = state.ui.settings.defaultIsoFiat
  const nativeToExchangeRatio = getExchangeDenomination(state, currencyCode)
  const exchangeAmount = convertNativeToExchange(nativeToExchangeRatio)(nativeAmount)
  const fiatAmount = convertCurrency(account, currencyCode, isoFiatCurrencyCode, exchangeAmount)
  const pinIsRequired = fiatAmount >= pinAmount

  return pinIsRequired
}
