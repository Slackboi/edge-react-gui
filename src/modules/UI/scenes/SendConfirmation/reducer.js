// @flow

import { combineReducers } from 'redux'
import type { EdgeSpendInfo, EdgeTransaction } from 'edge-core-js'
import { add } from 'biggystring'
import type { Action } from '../../../ReduxTypes.js'
import * as ACTION from './action'

export type SpendOptions = {
  lock: boolean,
  sign: boolean,
  broadcast: boolean,
  save: boolean
}

export type SendConfirmationState = {
  destination: string,
  error: Error | null,
  nativeAmount: string,
  pending: boolean,
  pin: string,
  spendInfo: EdgeSpendInfo | null,
  spendOptions: SpendOptions | null,
  transaction: EdgeTransaction | null
}

export const initialState = {
  destination: '',
  error: null,
  isEditable: true,
  nativeAmount: '0',
  pending: false,
  pin: '',
  spendInfo: null,
  transaction: null
}

export const spendInfo = (state: EdgeSpendInfo | null = null, action: Action) => {
  switch (action.type) {
    case ACTION.NEW_NATIVE_AMOUNT: {
      const { nativeAmount } = action.data
      const [first, ...rest] = state.spendTargets
      return {
        ...state,
        spendTargets: [{ ...first, nativeAmount }, ...rest]
      }
    }
    case ACTION.NEW_UNIQUE_IDENTIFIER: {
      const { uniqueIdentifier } = action.data
      const [first, ...rest] = state.spendTargets
      return {
        ...state,
        spendTargets: [
          {
            ...first,
            otherParams: {
              ...first.otherParams,
              uniqueIdentifier
            }
          },
          ...rest
        ]
      }
    }
    case ACTION.NEW_NETWORK_FEES: {
      const { customNetworkFee, networkFeeOption } = action.data
      return {
        ...state,
        networkFeeOption,
        customNetworkFee
      }
    }
    case ACTION.NEW_SPEND_REQUEST: {
      return action.data.spendInfo
    }
    case ACTION.RESET: {
      return null
    }
    default:
      return state
  }
}

export const error = (state: Error | null = null, action: Action) => {
  switch (action.type) {
    case ACTION.NEW_ERROR: {
      return action.data.error
    }
    case ACTION.NEW_SPEND_REQUEST:
    case ACTION.NEW_TRANSACTION:
    case ACTION.RESET: {
      return null
    }
    default:
      return state
  }
}

export const isEditable = (state: boolean = true, action: Action) => {
  switch (action.type) {
    case ACTION.NEW_SPEND_REQUEST: {
      return !action.data.options.locked
    }
    case ACTION.RESET: {
      true
    }
    default:
      return state
  }
}

export const pending = (state: boolean = false, action: Action) => {
  switch (action.type) {
    case ACTION.NEW_SPEND_REQUEST: {
      return action.data.options.sign
    }
    case ACTION.NEW_ERROR:
    case ACTION.RESET: {
      return false
    }
    default:
      return state
  }
}

export const pin = (state: string = '', action: Action) => {
  switch (action.type) {
    case ACTION.NEW_PIN: {
      return action.data.pin
    }
    case ACTION.NEW_ERROR:
    case ACTION.RESET: {
      return ''
    }
    default:
      return state
  }
}

export const authRequired = (state: 'pin' | 'none' = 'none', action: Action) => {
  switch (action.type) {
    case ACTION.NEW_SPEND_REQUEST: {
      return action.data.options.authRequired
    }
    case ACTION.RESET: {
      return false
    }
    default:
      return state
  }
}

export const transaction = (state: EdgeTransaction | null = null, action: Action) => {
  switch (action.type) {
    case ACTION.NEW_TRANSACTION: {
      return action.data.transaction
    }
    case ACTION.NEW_SPEND_REQUEST:
    case ACTION.RESET: {
      return null
    }
    default:
      return state
  }
}

export const nativeAmount = (state: string = '0', action: Action) => {
  switch (action.type) {
    case ACTION.NEW_NATIVE_AMOUNT: {
      return action.data.nativeAmount
    }
    case ACTION.NEW_SPEND_REQUEST: {
      return action.data.spendInfo.spendTargets.reduce((sum, { nativeAmount }) => {
        return add(sum, nativeAmount)
      }, '0')
    }
    case ACTION.RESET: {
      return ''
    }
    default:
      return state
  }
}

export const destination = (state: string = '', action: Action) => {
  switch (action.type) {
    case ACTION.NEW_SPEND_REQUEST: {
      return action.data.spendInfo.metadata.name
    }
    case ACTION.RESET: {
      return ''
    }
    default:
      return state
  }
}

export const resetSlider = (state: boolea = false, action: Action) => {
  switch (action.type) {
    case ACTION.NEW_ERROR:
    case ACTION.RESET: {
      return true
    }
    default:
      return state
  }
}

export const sliderDisabled = (state: boolea = false, action: Action) => {
  switch (action.type) {
    case ACTION.NEW_TRANSACTION: {
      return false
    }
    case ACTION.NEW_SPEND_REQUEST:
    case ACTION.RESET: {
      return true
    }
    default:
      return state
  }
}

export const sendConfirmation = combineReducers({
  authRequired,
  error,
  isEditable,
  nativeAmount,
  pending,
  pin,
  spendInfo,
  transaction,
  destination,
  resetSlider
})

export default sendConfirmation
