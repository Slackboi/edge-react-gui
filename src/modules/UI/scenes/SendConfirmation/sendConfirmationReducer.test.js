/* eslint-disable flowtype/require-valid-file-annotation */

/* globals test expect */

import { sendConfirmation } from './reducer.js'
import { initialState } from './selectors.js'
import { newPin, reset, updateSpendPending, updatePaymentProtocolTransaction, makeSpendFailed, updateTransaction, newSpendInfo } from './action.js'

describe('sendConfirmation reducer', () => {
  test('initialState', () => {
    const expected = initialState
    const actual = sendConfirmation(undefined, {})

    expect(actual).toMatchSnapshot()
  })

  test('reset', () => {
    const action = reset()
    const actual = sendConfirmation(undefined, action)

    expect(actual).toMatchSnapshot()
  })

  describe('isEditable', () => {
    test('UPDATE_PAYMENT_PROTOCOL_TRANSACTION', () => {
      const edgeTransaction = { id: '123', nativeAmount: '123' }
      const action = updatePaymentProtocolTransaction(edgeTransaction)
      const actual = sendConfirmation(initialState, action)

      expect(actual).toMatchSnapshot()
    })

    test('MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED', () => {
      const error = new Error()
      const action = makeSpendFailed(error)
      const actual = sendConfirmation(initialState, action)

      expect(actual).toMatchSnapshot()
    })
  })

  describe('error', () => {
    test('UPDATE_TRANSACTION', () => {
      const error = new Error()
      const action = updateTransaction(null, null, null, error)
      const actual = sendConfirmation(initialState, action)

      expect(actual).toMatchSnapshot()
    })

    test('MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED', () => {
      const error = new Error()
      const action = makeSpendFailed(error)
      const actual = sendConfirmation(initialState, action)

      expect(actual).toMatchSnapshot()
    })
  })

  test('pin', () => {
    const action = newPin('1234')
    const actual = sendConfirmation(initialState, action)

    expect(actual).toMatchSnapshot()
  })

  test('pin', () => {
    const action = newPin('1234')
    const actual = sendConfirmation(initialState, action)

    expect(actual).toMatchSnapshot()
  })

  test('pending', () => {
    const action = updateSpendPending(true)
    const actual = sendConfirmation(initialState, action)

    expect(actual).toMatchSnapshot()
  })
})
