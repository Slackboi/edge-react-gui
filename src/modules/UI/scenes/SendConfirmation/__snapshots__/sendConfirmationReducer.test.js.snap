// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendConfirmation reducer error MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED 1`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": [Error],
  "forceUpdateGuiCounter": 0,
  "isEditable": false,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
}
`;

exports[`sendConfirmation reducer error UPDATE_TRANSACTION 1`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": [Error],
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": null,
}
`;

exports[`sendConfirmation reducer initialState 1`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
}
`;

exports[`sendConfirmation reducer isEditable MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED 1`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": [Error],
  "forceUpdateGuiCounter": 0,
  "isEditable": false,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
}
`;

exports[`sendConfirmation reducer isEditable UPDATE_PAYMENT_PROTOCOL_TRANSACTION 1`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": false,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "id": "123",
    "nativeAmount": "123",
  },
}
`;

exports[`sendConfirmation reducer pending 1`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": true,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
}
`;

exports[`sendConfirmation reducer pin 1`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "1234",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
}
`;

exports[`sendConfirmation reducer pin 2`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "1234",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
}
`;

exports[`sendConfirmation reducer reset 1`] = `
Object {
  "authRequired": "none",
  "destination": "",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
}
`;
